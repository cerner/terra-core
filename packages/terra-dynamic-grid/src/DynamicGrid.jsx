import React from 'react';
import { StyleSheet, css } from 'aphrodite';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { grid } from './styles';
import Region from './Region';
import styles from './DynamicGrid.module.scss';

const templateShape = {
  /**
  * The column definitions of the grid. See https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns.
  * For IE support, verify syntax is supported here https://msdn.microsoft.com/en-us/library/hh772246(v=vs.85).aspx.
  */
  'grid-template-columns': PropTypes.string,
  /**
  * The row definitions of the grid. See https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-rows.
  * For IE support, verify syntax is supported here https://msdn.microsoft.com/en-us/library/hh772258(v=vs.85).aspx.
  */
  'grid-template-rows': PropTypes.string,
  /**
  * The gap to place between the columns and rows. See https://developer.mozilla.org/en-US/docs/Web/CSS/grid-gap.
  */
  'grid-gap': PropTypes.string,
  /**
  * Additional CSS properties to apply to the grid.
  */
  // eslint-disable-next-line react/forbid-prop-types
  style: PropTypes.object,
};


const propTypes = {
  /**
  * The child Region components that make up the grid.
  */
  children: PropTypes.node,

  /*
  * The default grid layout template to use.
  */
  defaultTemplate: PropTypes.shape(templateShape),

  /*
  *  The template to be used at tiny breakpoints
  */
  tiny: PropTypes.shape(templateShape),

  /*
  *  The template to be used at small breakpoints
  */
  small: PropTypes.shape(templateShape),

  /*
  *  The template to be used at medium breakpoints
  */
  medium: PropTypes.shape(templateShape),

  /*
  *  The template to be used at large breakpoints
  */
  large: PropTypes.shape(templateShape),

  /*
  *  The template to be used at huge breakpoints
  */
  huge: PropTypes.shape(templateShape),
};


const defaultProps = {
  defaultTemplate: {},
  tiny: {},
  small: {},
  medium: {},
  large: {},
  huge: {},
};

const DynamicGrid = ({
  defaultTemplate,
  tiny,
  small,
  medium,
  large,
  huge,
  children,
  ...customProps
}) => {
  const media = (breakpoint, props) => (Object.keys(props).length
    ? { [`@media screen and (min-width: ${breakpoint})`]: grid(props) }
    : {});

  const stylesheet = StyleSheet.create({
    grid: {
      ...grid(defaultTemplate),
      ...media(styles.tiny, tiny),
      ...media(styles.small, small),
      ...media(styles.medium, medium),
      ...media(styles.large, large),
      ...media(styles.huge, huge),
    },
  });

  const gridClasses = classNames(customProps.className, css(stylesheet.grid));
  return (<div {...customProps} className={gridClasses}>{children}</div>);
};

DynamicGrid.propTypes = propTypes;
DynamicGrid.defaultProps = defaultProps;
DynamicGrid.Region = Region;

export default DynamicGrid;
