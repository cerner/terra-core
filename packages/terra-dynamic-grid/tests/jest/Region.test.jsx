import React from 'react';
import DynamicGrid from '../../src/DynamicGrid';

describe('DynamicGrid', () => {
  const defaultRender = <DynamicGrid.Region name="foo" />;

  // Snapshot Tests
  it('should render a default component', () => {
    const region = shallow(defaultRender);
    expect(region).toMatchSnapshot();
  });

  it('should render a Grid with merged attributes', () => {
    const region = shallow(<DynamicGrid.Region className="TestClass" data-mock="data" style={{ height: '100px' }} />);
    expect(region).toMatchSnapshot();
  });

  // Attributes
  it('should merge classes passed in with attributes', () => {
    const region = shallow(<DynamicGrid.Region className="TestClass" />);
    expect(region.prop('className')).toContain('TestClass');
  });
});
