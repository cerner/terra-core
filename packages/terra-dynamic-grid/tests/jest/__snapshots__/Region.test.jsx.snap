// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicGrid should render a Grid with merged attributes 1`] = `
<div
  className="TestClass region_3hmsj"
  data-mock="data"
  style={
    Object {
      "height": "100px",
    }
  }
>
  <div
    className="region-child-container"
  />
</div>
`;

exports[`DynamicGrid should render a default component 1`] = `
<div
  className="region_3hmsj"
  name="foo"
>
  <div
    className="region-child-container"
  />
</div>
`;
